ctapipe/reco/sklearn.py

```diff
@@ -0,0 +1,590 @@
+
+    def fit(self, key, table):
+        """
+        Create and fit a new model for ``key`` using the data in ``table``.
+        """
+        self._models[key] = self._new_model()
+
+        X, valid = self._table_to_X(table)
+        y = self._table_to_y(table, mask=valid)
+        self.unit = table[self.target].unit
+        self._models[key].fit(X, y)
+
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        state["_trait_values"]["parent"] = None
+        state["_trait_notifiers"] = {}
+        return state
+
+
+class SKLearnRegressionReconstructor(SKLearnReconstructor):
+    """
+    Base class for regression tasks
+    """
+
+    model_cls = Enum(
+        SUPPORTED_REGRESSORS.keys(),
+        default_value=None,
+        allow_none=True,
+        help="Which scikit-learn regression model to use.",
+    ).tag(config=True)
+
+    log_target = Bool(
+        default_value=False,
+        help="If True, the model is trained to predict the natural logarithm.",
+    ).tag(config=True)
+
+    def _predict(self, key, table):
+        if key not in self._models:
+            raise KeyError(
+                f"No model available for key {key},"
+                f" available models: {self._models.keys()}"
+            )
+        X, valid = self._table_to_X(table)
+        n_outputs = getattr(self._models[key], "n_outputs_", 1)
+
+        if n_outputs > 1:
+            shape = (len(table), n_outputs)
+        else:
+            shape = (len(table),)
+
+        prediction = np.full(shape, np.nan)
+        if np.any(valid):
+            valid_predictions = self._models[key].predict(X)
+
+            if self.log_target:
+                prediction[valid] = np.exp(valid_predictions)
+            else:
+                prediction[valid] = valid_predictions
+
+        if self.unit is not None:
+            prediction = u.Quantity(prediction, self.unit, copy=False)
+
+        return prediction, valid
+
+    def _table_to_y(self, table, mask=None):
+        y = super()._table_to_y(table, mask=mask)
+
+        if self.log_target:
+            if np.any(y <= 0):
+                raise ValueError("y contains negative values, cannot apply log")
+
+            return np.log(y)
+        return y
+
+
+class SKLearnClassficationReconstructor(SKLearnReconstructor):
+    """
+    Base class for classification tasks
+    """
+
+    model_cls = Enum(
+        SUPPORTED_CLASSIFIERS.keys(),
+        default_value=None,
+        allow_none=True,
+        help="Which scikit-learn regression model to use.",
+    ).tag(config=True)
+
+    invalid_class = Integer(
+        default_value=-1, help="The label to fill in case no prediction could be made"
+    ).tag(config=True)
+
+    positive_class = Integer(
+        default_value=1,
+        help=(
+            "The label value of the positive class."
+            " ``prediction`` values close to 1.0 mean the event"
+            " belonged likely to this class."
```

Comment:
This is actually true even for the binary case with only one output label, but at some point we should note how the output differs between binary and multi class problems.